replica-node/src/replica_node.rs

```diff
@@ -0,0 +1,319 @@
+    pub ledger_path: PathBuf,
+    pub snapshot_output_dir: PathBuf,
+    pub snapshot_path: PathBuf,
+    pub account_paths: Vec<PathBuf>,
+    pub snapshot_info: (Slot, Hash),
+    pub cluster_info: Arc<ClusterInfo>,
+    pub rpc_config: JsonRpcConfig,
+    pub snapshot_config: Option<SnapshotConfig>,
+    pub pubsub_config: PubSubConfig,
+    pub account_indexes: AccountSecondaryIndexes,
+    pub accounts_db_caching_enabled: bool,
+    pub replica_exit: Arc<RwLock<Exit>>,
+}
+
+impl Default for ReplicaNodeConfig {
+    fn default() -> Self {
+        Self {
+            rpc_source_addr: SocketAddr::from(([127, 0, 0, 1], 8001)),
+            rpc_addr: SocketAddr::from(([127, 0, 0, 1], 8001)),
+            rpc_pubsub_addr: SocketAddr::from(([127, 0, 0, 1], 8001)),
+            ledger_path: PathBuf::default(),
+            snapshot_output_dir: PathBuf::default(),
+            snapshot_path: PathBuf::default(),
+            account_paths: vec![],
+            snapshot_info: (0, Hash::default()),
+            cluster_info: Arc::new(ClusterInfo::default()),
+            rpc_config: JsonRpcConfig::default(),
+            snapshot_config: None,
+            pubsub_config: PubSubConfig::default(),
+            account_indexes: AccountSecondaryIndexes::default(),
+            accounts_db_caching_enabled: false,
+            replica_exit: Arc::new(RwLock::new(Exit::default())),
+        }
+    }
+}
+
+pub struct ReplicaNode {
+    json_rpc_service: Option<JsonRpcService>,
+    pubsub_service: Option<PubSubService>,
+    optimistically_confirmed_bank_tracker: Option<OptimisticallyConfirmedBankTracker>,
+}
+
+// Struct maintaining information about banks
+struct ReplicaBankInfo {
+    bank_forks: Arc<RwLock<BankForks>>,
+    optimistically_confirmed_bank: Arc<RwLock<OptimisticallyConfirmedBank>>,
+    leader_schedule_cache: Arc<LeaderScheduleCache>,
+    block_commitment_cache: Arc<RwLock<BlockCommitmentCache>>,
+}
+
+// Initialize the replica by downloading snapshot from the peer, initialize
+// the BankForks, OptimisticallyConfirmedBank, LeaderScheduleCache and
+// BlockCommitmentCache and return the info wrapped as ReplicaBankInfo.
+fn initialize_from_snapshot(
+    replica_config: &ReplicaNodeConfig,
+    snapshot_config: &SnapshotConfig,
+    genesis_config: &GenesisConfig,
+) -> ReplicaBankInfo {
+    info!(
+        "Downloading snapshot from the peer into {:?}",
+        replica_config.snapshot_output_dir
+    );
+
+    download_snapshot(
+        &replica_config.rpc_source_addr,
+        &replica_config.snapshot_output_dir,
+        replica_config.snapshot_info,
+        false,
+        snapshot_config.maximum_snapshots_to_retain,
+        &mut None,
+    )
+    .unwrap();
+
+    fs::create_dir_all(&snapshot_config.snapshot_path).expect("Couldn't create snapshot directory");
+
+    let archive_info = snapshot_utils::get_highest_full_snapshot_archive_info(
+        replica_config.snapshot_output_dir.to_path_buf(),
+    )
```

Comment:
nit: Since `replica_config.snapshot_output_dir` is a `PathBuf`, you should be able to pass this as a reference to `get_highest_full_snapshot_archive_info()`. i.e.
```suggestion
    let archive_info = snapshot_utils::get_highest_full_snapshot_archive_info(
        &replica_config.snapshot_output_dir,
    )
```